ktor-server/ktor-server-plugins/ktor-server-metrics/jvm/src/io/ktor/server/metrics/dropwizard/DropwizardMetrics.kt

```diff
@@ -10,133 +10,113 @@ import com.codahale.metrics.jvm.*
+        application.intercept(ApplicationCallPipeline.Monitoring) {
+            try {
+                proceed()
+            } finally {
+                handler(call)
+            }
         }
     }
+}
 
-    private data class CallMeasure(val timer: Timer.Context)
+public val DropwizardMetrics: ApplicationPlugin<Application, DropwizardMetricsConfig, PluginInstance> =
+    createApplicationPlugin("DropwizardMetrics", ::DropwizardMetricsConfig) {
+        val duration = pluginConfig.registry.timer(name(pluginConfig.baseName, "duration"))
+        val active = pluginConfig.registry.counter(name(pluginConfig.baseName, "active"))
+        val exceptions = pluginConfig.registry.meter(name(pluginConfig.baseName, "exceptions"))
+        val httpStatus = ConcurrentHashMap<Int, Meter>()
+
+        if (pluginConfig.registerJvmMetricSets) {
+            listOf<Pair<String, () -> Metric>>(
+                "jvm.memory" to ::MemoryUsageGaugeSet,
+                "jvm.garbage" to ::GarbageCollectorMetricSet,
+                "jvm.threads" to ::ThreadStatesGaugeSet,
+                "jvm.files" to ::FileDescriptorRatioGauge,
+                "jvm.attributes" to ::JvmAttributeGaugeSet
+            ).filter { (name, _) ->
+                !pluginConfig.registry.names.any { existingName -> existingName.startsWith(name) }
+            }.forEach { (name, metric) -> pluginConfig.registry.register(name, metric()) }
+        }
 
-    private val measureKey = AttributeKey<CallMeasure>("metrics")
+        on(CallFailed) { _, _ ->
+            exceptions.mark()
+        }
 
-    private fun before(call: ApplicationCall) {
-        active.inc()
-        call.attributes.put(measureKey, CallMeasure(duration.time()))
-    }
+        on(CallRouted) { call ->
+            val name = call.route.toString()
+            val meter = pluginConfig.registry.meter(name(pluginConfig.baseName, name, "meter"))
+            val timer = pluginConfig.registry.timer(name(pluginConfig.baseName, name, "timer"))
+            meter.mark()
+            val context = timer.time()
+            call.attributes.put(
+                routingMetricsKey,
+                RoutingMetrics(name, context)
+            )
+        }
 
-    private fun after(call: ApplicationCall) {
-        active.dec()
-        val meter = httpStatus.computeIfAbsent(call.response.status()?.value ?: 0) {
-            registry.meter(name(baseName, "status", it.toString()))
+        on(RoutedCallProcessed) { call ->
+            val routingMetrics = call.attributes.take(routingMetricsKey)
+            val status = call.response.status()?.value ?: 0
+            val statusMeter =
+                pluginConfig.registry.meter(name(pluginConfig.baseName, routingMetrics.name, status.toString()))
+            statusMeter.mark()
+            routingMetrics.context.stop()
         }
-        meter.mark()
-        call.attributes.getOrNull(measureKey)?.apply {
-            timer.stop()
+
+        on(Monitoring) { call ->
```

Comment:
It's not clear when monitoring is happening, consider to rename it

Reply:
Would `BeforeCall` work here? 
Or maybe `TracingCall`